src/workerd/jsg/filesystem.h

```diff
@@ -0,0 +1,244 @@
+
+  size_t write(Lock& js, size_t offset, kj::StringPtr data) {
+    return write(js, offset, data.asBytes());
+  }
+
+  // Zero out the contents of the file.
+  virtual void zero(Lock& js, kj::Maybe<size_t> offset = kj::none) = 0;
+
+  // Copies this file into a new orphaned file that is writable.
+  virtual kj::Own<File> cloneAsWritable(Lock& js) const = 0;
+
+  // Resize the file allocation. Note that this is potentially an expensive
+  // operation as it requires allocating a new internal buffer and copying
+  // the data. If the size is smaller than the current size, the contents of
+  // the fill will be truncated. If the size is larger than the current size,
+  // the new contents of the file will be filled with zeroes.
+  virtual void resize(Lock& js, size_t size) = 0;
+
+  // Creates a new readable/writable in-memory file. This file will not be
+  // initially included in a directory. To add it to a directory, use the
+  // transfer method. If the file is not transferred, it will be deleted
+  // when the handle is dropped. If size is given, the file will be initially
+  // filled with zeroes up to the given size, otherwise the file will be empty.
+  // The contents of the file will be tracked and counted towards the isolate
+  // external memory usage.
+  // If size is not given, the file will be empty. If the intent is to perform
+  // multiple writes to the file, it is recommended to specify a size up front
+  // or to resize the file as appropriate to account for the expected writes.
+  // This will avoid each individual write causing a reallocation of the
+  // internal buffer to accommodate the new data.
+  static kj::Own<File> newWritable(Lock& js, kj::Maybe<size_t> size = kj::none);
+
+  // Creates a new readable in-memory file wrapping the given data. The file
+  // does not take ownership of the data and the data must remain valid for the
+  // lifetime of the file. The file will be read-only. It will not be initially
+  // included in a directory. The contents of the file will not be tracked and
+  // will not count towards the isolate external memory usage.
+  static kj::Own<File> newReadable(kj::ArrayPtr<const kj::byte> data);
+};
+
+// A directory in the virtual file system. If the directory is read-only,
+// then the mutation methods will throw an exception.
+class Directory: public Node {
+ public:
+  kj::Own<Directory> addRef(jsg::Lock& js) {
+    return kj::addRef(*this);
+  }
+
+  // Return the number of entries in this directory. If typeFilter is provided,
+  // only entries matching the given type will be counted.
+  virtual size_t count(Lock& js, kj::Maybe<Type> typeFilter = kj::none) const = 0;
+
+  // Iterate over the entries in this directory. If typeFilter is provided,
+  // only entries matching the given type will be returned.
+  virtual void forEach(Lock& js,
+      kj::FunctionParam<bool(Lock& js, kj::StringPtr, const Stat stat)> fn,
```

Comment:
Why do we need this to receive a callback?
Can't we just as well turn Directory into an iterator and have a C++ idiomatic interface for iterating it?

Reply:
Should be as simple as begin() {return entries.begin()} end(){return entries.end()}

Reply:
Other than the fact that I despise implementing to the C++ idiomatic iterator API? No reason I suppose ;-)